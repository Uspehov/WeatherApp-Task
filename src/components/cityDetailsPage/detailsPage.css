.detailsPage-container {
    text-align: center;
    color: white;
    width: 100%;
    height: 100%;
    position:absolute; 
    top:0; 
    left:0;
    overflow-x: scroll;
}

.Moscow {
    background-size: cover;
    background: linear-gradient(rgba(0, 0, 0, .5), rgba(0, 0, 0, .5)),url('https://www.aeroflot.ru/media/aflfiles/travelguide/ru/mow/mow_1.jpg') no-repeat center center fixed;
}

.Rostov-on-Don{
    background-size: cover;
    background: linear-gradient(rgba(0, 0, 0, .5), rgba(0, 0, 0, .5)),url('https://i.ytimg.com/vi/M0apxj0oRYM/maxresdefault.jpg') no-repeat center center fixed;
}

.SaintPetersburg {
    background-size: cover;
    background: linear-gradient(rgba(0, 0, 0, .5), rgba(0, 0, 0, .5)),url('https://gazeta.a42.ru/uploads/f21/f211cc50-62e8-11e8-bc25-d3fc31daadda.jpg') no-repeat center center fixed;
}

.detailsPage-loader {
    display: flex;
    align-items: center;
    justify-content: center;
    position:absolute;
    width: 100%;
    height: 100%;
    background-color: cornsilk;
    top:0; 
    left:0;
}

.error {
    background-color: grey;
    display: flex;
    flex-direction: column;
}

.white {
    color: white;
}