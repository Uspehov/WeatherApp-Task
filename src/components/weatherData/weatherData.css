.weatherData-container {
    font-size: 25px; 
}

.weatherData-subtitle {
    margin: 15px;
}

.weatherData-columnContainer {
    margin: auto;
    width: 35%;
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-top: 40px;
}

.weatherData-column {
    display: flex;
    flex-direction: column;
}

@media (max-width: 750px) {
    .weatherData-columnContainer {width: 65%;}
}
