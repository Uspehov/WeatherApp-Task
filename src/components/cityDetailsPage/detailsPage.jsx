import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Loader from 'react-loader-spinner';
import { connect } from "react-redux";
import { getWeather } from '../../js/actions';
import { BackToMainButton } from '../buttons/backToMain/backToMain';
import { WeatherData } from '../weatherData/weatherData';
import getWeatherDataSelector from '../../js/selectors/index';
import { withRouter } from "react-router";
import './detailsPage.css';

export class DetailsPage extends Component {
    constructor(props) {
        super(props);
        this.state = { isLoading: true }
    }

    componentWillMount() {
        this.props.getData(this.props.cityName);
    }

    componentWillReceiveProps(nextProps) {
        this.setState({ isLoading: nextProps.isLoading });
    }

    render() {
        return (
            <div>
                {!this.state.isLoading ?
                    <div>
                        {!this.props.isError && this.props.data.weather ?
                            <div className={`detailsPage-container ${this.props.cityName.replace(/\s/g, '')}`}>
                                <WeatherData
                                    data={this.props.data}
                                    cityName={this.props.cityName}
                                    isLoading={this.props.isLoading}
                                />
                            </div>
                            :
                            <div className='detailsPage-loader error'>
                                <h1 className='white'>Ooops, something went wrong! Please try again later!</h1>
                                <BackToMainButton />
                            </div>
                        }
                    </div>
                    :
                    <div className='detailsPage-loader'>
                        <Loader
                            visible={this.state.isLoading}
                            type="Puff"
                            color="#00BFFF"
                            height={100}
                            width={100}
                        />
                    </div>
                }
            </div>
        );
    }
}

const makeMapStateToProps = () => {
    const uniqueWeatherData = getWeatherDataSelector();
    const mapStateToProps = (state, props) => {
        return {
            isLoading: state.isLoading,
            isError: state.isError,
            data: uniqueWeatherData(state, props),
        }
    }
    return mapStateToProps;
}

const mapDispatchToProps = (dispatch) => {
    return {
        getData: (cityName) => dispatch(getWeather(cityName))
    };
};

DetailsPage.propTypes = {
    isLoading: PropTypes.bool,
    isError: PropTypes.bool,
    data: PropTypes.object,
    cityName: PropTypes.string,
    getData: PropTypes.func
}

export default withRouter(connect(makeMapStateToProps, mapDispatchToProps)(DetailsPage));