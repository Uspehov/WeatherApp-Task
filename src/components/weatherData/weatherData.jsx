import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { BackToMainButton } from '../buttons/backToMain/backToMain';
import { weatherValues } from '../../js/constants/weatherValues';
import './weatherData.css';

export class WeatherData extends Component {

    render() {
        const data = this.props.data;
        return (
            <div className='weatherData-container'>
                <h1>{this.props.cityName}</h1>
                <div >
                    <h3 className='weatherData-subtitle'>{data.weather[0].main}</h3>
                    <span>{data.weather[0].description}</span>
                    <div className='weatherData-columnContainer'>
                        <div className='weatherData-column'>{Object.keys(data.main).map((item) => <div key={item}>{item.charAt(0).toLocaleUpperCase() + item.slice(1).replace(/_/, '-')}</div>)}</div>
                        <div className='weatherData-column'>{Object.values(data.main).map((item, index) => <div key={item}>{item} {weatherValues[index]}</div>)} </div>
                    </div>
                    <h3 className='weatherData-subtitle'>Wind</h3>
                    <div className='weatherData-columnContainer'>
                        <div className='weatherData-column'>{Object.keys(data.wind)[0].charAt(0).toLocaleUpperCase() + Object.keys(data.wind)[0].slice(1)}</div>
                        <div className='weatherData-column'>{Object.values(data.wind)[0]} m/s</div>
                    </div>
                </div>
                <BackToMainButton />
            </div>
        );
    }
}

WeatherData.propTypes = {
    cityName: PropTypes.string,
    data: PropTypes.object
}