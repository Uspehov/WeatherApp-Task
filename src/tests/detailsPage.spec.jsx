import React from 'react';
import  { configure, shallow }  from 'enzyme';
import  { DetailsPage }  from '../components/cityDetailsPage/detailsPage';
import { WeatherData } from '../components/weatherData/weatherData';
import Loader from 'react-loader-spinner';
import Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

describe('Details  Page', () => {
    const detailsPageProps = {
        isLoading: false,
        isError: false,
        data: {
            weather: [{
                data: 'sun',
                main: 'clear',
                wind: '4',
                description: 'good day'
            }],
            main: {
                humidity: 79,
                pressure: 1030,
                temp: 268.03,
                temp_max: 269.26,
                temp_min: 267.04
            },
            wind: {speed: 2, deg: 170}
        },
        getData: jest.fn(),
        cityName: 'Moscow'
    };

    it('should render', () => {
        const wrapper = shallow(<DetailsPage {...detailsPageProps}/>);
        wrapper.setState({ isLoading: false });
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('Loader should show if Pending', () => {
        const wrapper = shallow(<DetailsPage {...detailsPageProps} isLoading={true} />);
        expect(wrapper.find('.detailsPage-loader').length).toBe(1);
        expect(wrapper.find(Loader).length).toBe(1);
    });

    it('if error, proper messedge should appear', () => {
        const wrapper = shallow(<DetailsPage {...detailsPageProps} isError={true}/>);
        wrapper.setState({ isLoading: false });
        expect(wrapper.find('.error').length).toBe(1);
    });

    it('if data is missing, error messedge should appear', () => {
        const wrapper = shallow(<DetailsPage {...detailsPageProps} isLoading={true} />);
        expect(wrapper.find('.detailsPage-loader').length).toBe(1);
        expect(wrapper.find(Loader).length).toBe(1);
    });

    it('if everything is normal, should render WeatherData component', () => {
        const wrapper = shallow(<DetailsPage {...detailsPageProps}/>);
        wrapper.setState({ isLoading: false });
        expect(wrapper.find(WeatherData).length).toBe(1);
    })
});