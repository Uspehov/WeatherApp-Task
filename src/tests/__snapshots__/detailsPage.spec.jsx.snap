// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details  Page should render 1`] = `
"<div>
  <div>
    <div className=\\"detailsPage-container Moscow\\">
      <WeatherData data={{...}} cityName=\\"Moscow\\" isLoading={false} />
    </div>
  </div>
</div>"
`;
